include/vcpkg/base/parallel-algorithms.h

```diff
@@ -1,85 +1,163 @@
+    struct WorkCallbackContext
     {
-        const size_t thread_count = static_cast<size_t>(get_concurrency());
-        const size_t num_threads = std::max(static_cast<size_t>(1), std::min(thread_count, work_count));
+        F work;
+        size_t work_count;
+        std::atomic<size_t> next_offset;
 
-        std::vector<std::future<void>> workers;
-        workers.reserve(num_threads - 1);
+        WorkCallbackContext(F init_f, size_t work_count) : work(init_f), work_count(work_count), next_offset(0) { }
 
-        for (size_t i = 0; i < num_threads - 1; ++i)
+        void run()
         {
-            workers.emplace_back(std::async(std::launch::async | std::launch::deferred, [&work]() { work(); }));
+            auto offset = next_offset.load(std::memory_order_relaxed);
+            while (offset < work_count)
+            {
+                if (!next_offset.compare_exchange_weak(offset, offset + 1, std::memory_order_relaxed))
+                {
+                    continue;
+                }
+
+                work(offset);
+            }
         }
-        work();
+    };
 
-        for (auto&& w : workers)
+#if defined(_WIN32)
+    struct PtpWork
+    {
+        PtpWork(_In_ PTP_WORK_CALLBACK pfnwk, _Inout_opt_ PVOID pv, _In_opt_ PTP_CALLBACK_ENVIRON pcbe)
+            : ptp_work(CreateThreadpoolWork(pfnwk, pv, pcbe))
         {
-            w.get();
         }
-    }
+        PtpWork(const PtpWork&) = delete;
+        PtpWork& operator=(const PtpWork&) = delete;
+        ~PtpWork()
+        {
+            if (ptp_work)
+            {
+                ::WaitForThreadpoolWorkCallbacks(ptp_work, TRUE);
+                ::CloseThreadpoolWork(ptp_work);
+            }
+        }
 
-    template<class Container, class F>
-    void parallel_for_each(Container&& c, F cb) noexcept
+        explicit operator bool() { return ptp_work != nullptr; }
+
+        void submit() { ::SubmitThreadpoolWork(ptp_work); }
+
+    private:
+        PTP_WORK ptp_work;
+    };
+
+    template<class F>
+    inline void execute_in_parallel(size_t work_count, F work) noexcept
     {
-        if (c.size() == 0)
+        if (work_count == 0)
         {
             return;
         }
-        if (c.size() == 1)
+
+        if (work_count == 1)
         {
-            cb(c[0]);
-            return;
+            work(size_t{});
         }
 
-        std::atomic_size_t next{0};
-
-        execute_in_parallel(c.size(), [&]() {
-            size_t i = 0;
-            while (i < c.size())
+        WorkCallbackContext<F> context{work, work_count};
+        PtpWork ptp_work([](PTP_CALLBACK_INSTANCE,
+                            void* context,
+                            PTP_WORK) noexcept { static_cast<WorkCallbackContext<F>*>(context)->run(); },
+                         &context,
+                         nullptr);
+        if (ptp_work)
+        {
+            auto max_threads = (std::min)(work_count, static_cast<size_t>(get_concurrency()));
```

Comment:
Since this is leveraging the thread pool, should this be larger than the hardware concurrency and letting the thread pool decide the appropriate oversubscription based on CPU vs IO?

Reply:
get_concurrency is `VCPKG_MAX_CONCURRENCY`, not hardware_concurrency. I'm inclined to do what the user said here unless we have concrete data saying we should try more.